internal/provider/resource_device.go

```diff
@@ -0,0 +1,216 @@
+	"github.com/paultyng/go-unifi/unifi"
+)
+
+func resourceDevice() *schema.Resource {
+	return &schema.Resource{
+		Description: "`unifi_device` manages a device of the network.\n\n" +
+			"Devices are adopted by the controller, so it may not be possible " +
+			"for this resource to be created through terraform.",
+
+		Create: resourceDeviceCreate,
+		Read:   resourceDeviceRead,
+		Update: resourceDeviceUpdate,
+		Delete: resourceDeviceDelete,
+		Importer: &schema.ResourceImporter{
+			State: importSiteAndID,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"id": {
+				Description: "The ID of the device.",
+				Type:        schema.TypeString,
+				Computed:    true,
+			},
+			"site": {
+				Description: "The name of the site to associate the device with.",
+				Type:        schema.TypeString,
+				Computed:    true,
+				Optional:    true,
+				ForceNew:    true,
+			},
+			"mac": {
+				Description:      "The MAC address of the device.",
+				Type:             schema.TypeString,
+				Required:         true,
+				ForceNew:         true,
+				DiffSuppressFunc: macDiffSuppressFunc,
+				ValidateFunc:     validation.StringMatch(macAddressRegexp, "Mac address is invalid"),
+			},
+			"name": {
+				Description: "The name of the device.",
+				Type:        schema.TypeString,
+				Optional:    true,
+			},
+			"disabled": {
+				Description: "Specifies whether this device should be disabled.",
+				Type:        schema.TypeBool,
+				Computed:    true,
+			},
+			"port_overrides": {
+				Description: "Settings overrides for specific switch ports.",
+				Type:        schema.TypeSet,
+				Optional:    true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"port_idx": {
+							Description: "Switch port number.",
+							Type:        schema.TypeInt,
+							Required:    true,
+						},
+						"name": {
+							Description: "Human-readable name of the port.",
+							Type:        schema.TypeString,
+							Optional:    true,
+						},
+						"port_profile_id": {
+							Description: "ID of the Port Profile used on this port.",
+							Type:        schema.TypeString,
+							Optional:    true,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func resourceDeviceCreate(d *schema.ResourceData, meta interface{}) error {
+	return errors.New("unifi_device can only be imported, not created")
+}
+
+func resourceDeviceUpdate(d *schema.ResourceData, meta interface{}) error {
+	c := meta.(*client)
+
+	site := d.Get("site").(string)
+	if site == "" {
+		site = c.site
+	}
+
+	req, err := resourceDeviceGetResourceData(d)
+	if err != nil {
+		return err
+	}
+
+	req.ID = d.Id()
+	req.SiteID = site
+
+	resp, err := c.c.UpdateDevice(context.TODO(), site, req)
+	if err != nil {
+		return err
+	}
+
+	return resourceDeviceSetResourceData(resp, d, site)
+}
+
+func resourceDeviceDelete(d *schema.ResourceData, meta interface{}) error {
+	return errors.New("Deleting unifi_device is not supported yet")
```

Comment:
Maybe this should just return a warning that its removed from state but not actually destroyed? You can do this in the `DeleteContext` method, using diagnostics. I can make this modification later as well, not a big deal.